manufacturer-toolkit/src/main/java/org/sdo/sct/mt/OnDieCertPath.java

```diff
@@ -0,0 +1,163 @@
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+
+@Service
+public class OnDieCertPath {
+
+  private CertificateFactory certFactory;
+
+  /**
+   * Constructor.
+   *
+   */
+  @Autowired
+  public OnDieCertPath() throws Exception {
+    certFactory = CertificateFactory.getInstance("X509", BouncyCastleSingleton.INSTANCE);
+  }
+
+  /**
+   * Builds a cert path for the OnDie device.
+   *
+   * @param b64DeviceCertChain the cert chain from ROM to leaf from the device
+   * @return a complete CA to leaf cert path
+   * @throws Exception if error
+   */
+  public CertPath buildCertPath(
+      String b64DeviceCertChain, OnDieCache onDieCertCache) throws Exception {
+
+    List<Certificate> deviceCertChain = new ArrayList<Certificate>();
+
+    // parse the device input chain into a list of x509 certificates
+    List<byte[]> certList = deserializeCertificateChain(Base64.decode(b64DeviceCertChain));
+    for (byte[] array : certList) {
+      InputStream is = new ByteArrayInputStream(array);
+      deviceCertChain.add(certFactory.generateCertificate(is));
+    }
+
+    // Based on the ROM certificate of the device, look up the issuing certificate
+    // and then each subsequent issuing certificate. Append them to the list
+    // to complete an end-to-end cert chain.
+    // leaf -> DAL -> Kernel -> ROM -> platform -> intermediate -> OnDie CA
+
+    // Start with the ROM cert, loop until we get to the CA
+    String certId  = getIssuingCertificate(deviceCertChain.get(0));
+    while (true) {
+      byte[] cert = onDieCertCache.getCertOrCrl(certId);
+      if (cert == null) {
+        throw new Exception(
+          "Could not find cert in OnDie cache: unable to build cert chain for device: " + certId);
+      }
+      ByteArrayInputStream is = new ByteArrayInputStream(cert);
+      X509Certificate x509Cert = (X509Certificate) certFactory.generateCertificate(is);
+      deviceCertChain.add(x509Cert);
+      is.close();
+
+      // if we are at the root CA then we have the complete chain
+      try {
+        certId = getIssuingCertificate(x509Cert);
+      } catch (Exception ex) {
+        break;
+      }
+    }
+
+    return certFactory.generateCertPath(deviceCertChain);
+  }
+
+  private String getIssuingCertificate(Certificate cert) throws Exception {
+    X509CertificateHolder certholder = new X509CertificateHolder(cert.getEncoded());
+    AuthorityInformationAccess aia =
+        AuthorityInformationAccess.fromExtensions(certholder.getExtensions());
+    if (aia == null) {
+      throw new Exception("AuthorityInformationAccess Extension missing from device certificate.");
+    }
+    AccessDescription[] descs = aia.getAccessDescriptions();
+    if (descs.length != 1) {
+      throw new Exception("Too many descriptions in AIA certificate extension: " + descs.length);
+    }
+    return descs[0].getAccessLocation().getName().toString();
+  }
+
+  /// <summary>
```

Comment:
Reformatting of the comment would be required.

Reply:
good catch. Checkstyle didn't complain because it was a private method.